lsp/server/src/server.ts

```diff
@@ -34,92 +34,85 @@ const documents: TextDocuments<TextDocument> = new TextDocuments(TextDocument);
+const execAsync = promisify(exec);
 
-	// Does the client support the `workspace/configuration` request?
-	// If not, we fall back using global settings.
-	hasConfigurationCapability = !!(
-		capabilities.workspace && !!capabilities.workspace.configuration
-	);
-	hasWorkspaceFolderCapability = !!(
-		capabilities.workspace && !!capabilities.workspace.workspaceFolders
-	);
-
-	const result: InitializeResult = {
-		capabilities: {
-			codeActionProvider: true,
-			textDocumentSync: TextDocumentSyncKind.Incremental,
-			executeCommandProvider: {
-				commands: ['sa4u.fix']
-			}
-		}
-	};
-	if (hasWorkspaceFolderCapability) {
-		result.capabilities.workspace = {
-			workspaceFolders: {
-				supported: true
-			}
-		};
+class SA4UConfig {
+	compilationDir: string
+	constructor(compilationDir: string) {
+		this.compilationDir = '/src/' + compilationDir;
 	}
-	return result;
-});
+}
 
-connection.onInitialized(() => {
-	
-	if (hasConfigurationCapability) {
-		// Register for all configuration changes.
-		connection.client.register(DidChangeConfigurationNotification.type, undefined);
-		connection.workspace.getConfiguration('SA4U').then((value) => {
-			if (value.messageDefinition) {
-				messDef = value.messageDefinition;
-			}
-		});
-	}
-	if (hasWorkspaceFolderCapability) {
-		connection.workspace.onDidChangeWorkspaceFolders(_event => {
-			connection.console.log('Workspace folder change event received.');
-		});
-	}
-	connection.workspace.getConfiguration();
+async function checkForDockerRunningAndInstalled() {
+	return await execAsync('docker info')
+		.then(() => {return Promise.resolve;}, () => {return Promise.reject('Please verify Docker is installed and Docker daemon is running.');});
+}
+
+async function startDockerContainer() {
 	connection.workspace.getWorkspaceFolders().then((folders:null|WorkspaceFolder[]|void) => {
 		if (folders) {
 			folders.forEach((folder) => {
 				dockerContainer(folder);
 			});
 		}
 	});
-	documents.all().forEach((doc)=>validateTextDocument(doc));
-});
-
-class SA4UConfig {
-	compilationDir: string
-	constructor(compilationDir: string) {
-		this.compilationDir = '/src/' + compilationDir;
-	}
 }
 
-// TODO: it would be nice to automatically discover the compile_commands.json file.
-const readFileAsync = promisify(readFile);
+async function stopDockerContainer() {
+	const output = await execAsync(`docker container ls -q -f name=sa4u_z3_server`)
+		.then((value) => {return value.stdout.split('\n');})
+		.catch((err) => {console.log(err); return [];});
+	await execAsync(`docker container stop ${output.join(' ')}`);
+}
 
-async function getSA4UConfig(projectRootDir: string) {
-	try {
-		const configJSON = JSON.parse((await readFileAsync(projectRootDir + "/" + ".sa4u.json")).toString());
-		return new SA4UConfig(configJSON['CompilationDir']);
-	} catch (err) {
-		console.warn('unable to read SA4U config: ' + err);
+async function getSA4UConfig() {
+	const readConfig = await promises.readFile(".sa4u.json")
```

Comment:
Can we use `try`/`catch` blocks instead of `then`/`catch` chaining?
Something like:

```
let configFileContents: String;
try {
    const configFileContents = (await promises.readFile('.sa4u.json')).toString();
} catch (error) {
    console.warn(...);
    return new SA4UConfig('');
}
```

A quick experiment:
```
const otherFunc = async (): Promise<number> => {
    throw new Error('I cannot make a value!');
}

const myFunc = async (): Promise<number> => {
    let x;
    try {
        x = await otherFunc();
    } catch (error) {
        console.log('I received an error: ' + error);
        return 0;
    }
    return x * 10;
}

myFunc().then((value: number) => console.log(value));
```
Indicates that this works. I like this approach more because we can get rid of the `instanceof` checks.

Reply:
Fixed